chore(docs): add sequencer-relayer doc to specs (#1126)
astriaorg/astria: specs/sequencer-relayer.md

# Sequencer-Relayer

## Overview

The sequencer-relayer, a.k.a relayer, is an application whose primary

> responsibility is to store sequenced rollup data on a remote data availability

Reviewer: ```suggestion
responsibility is to publish sequenced rollup data on a remote data availability
```

Author: Done in [045e82c](https://github.com/astriaorg/astria/pull/1126/commits/045e82c6be5afcd9e2fd53055e3d977e4037ecbf).

layer (DA layer).

The immediate source of the sequenced data is the sequencer blockchain network,
where rollups' data is sequenced and included in sequencer blocks, while the DA
layer is the Celestia blockchain.

## Application Logic

The application logic of the relayer is comprised of three main tasks: getting
the data from a sequencer node, transforming the data in preparation for

> storing, and putting the transformed data onto the Celestia network.

Reviewer: ```suggestion
publishing, and putting the transformed data onto the Celestia network.
```

### Getting the Data from a Sequencer Node

The relayer runs an endlessly repeating task, the "reader task", which does the
following:

1. polls the sequencer node for the latest (highest) sequencer block height at
  a fixed frequency (once per second currently)

> 1. upon learning of a new height, fetches the block at that height from the
>   sequencer node
> 1. upon receiving the sequencer block, forwards it to the transformation task

Reviewer: ```suggestion
2. upon learning of a new height, fetches the block at that height from the
  sequencer node
3. upon receiving the sequencer block, forwards it to the transformation task
```

Author: The numbers are rendered correctly in MarkDown and I tend to use all 1s since it's easier to keep the numbered list correct if you reorder, add or remove items.

Happy to change it if you prefer though - not a big deal :)

Reviewer: oh interesting, i didn't read the rendered markdown lol, no worries then

Note that currently the relayer only communicates with a single sequencer node,
and that this node is comprised of two separate processes, each providing their
own http servers. Consequently, the relayer queries two different endpoints
using two different protocols; CometBFT's ABCI `Info` for the block height via
JSONRPC, and the sequencer app's `GetSequencerBlock` via gRPC. This is likely to
improve in the near future, as work is under way to extend the sequencer app's
API to the extent that no direct communication with the CometBFT node will be
required.

Should the sequencer fail to fetch the latest sequencer block height, no action
is taken other than logging and recording the fact in metrics - polling

> continues at the same frequency under tha assumption that the error is a

Reviewer: ```suggestion
continues at the same frequency under the assumption that the error is a
```

Author: Well spotted! Fixed in [045e82c](https://github.com/astriaorg/astria/pull/1126/commits/045e82c6be5afcd9e2fd53055e3d977e4037ecbf).

On receiving each new sequencer block, the relayer transforms it (see below for
details) and tries to add it to the next batch for submission to the Celestia
node. If adding it would cause the batch size to exceed an upper limit
(currently 1MB), then it is temporarily cached instead and receiving further
sequencer blocks from the reader task is paused.

Note that the reader task itself isn't immediately stopped in this case. A
bounded channel is used to send sequencer blocks from the reader task to the
submitter task, currently with a limit of 128 blocks. Normally this channel will
have 0 or 1 blocks, but if the submitter task pauses consumption of the blocks,
the reader task will continue to send blocks into the channel until it's full,
at which point the reader task also pauses until the channel has capacity to
send again.

The transformation steps on receiving a new sequencer block are as follows:

1. The rollup transactions are extracted into a separate collection of
  `SubmittedRollupData`s (one per rollup) where each includes the sequencer
  block hash. The rest of the block's contents and the list of included rollup
  IDs is moved to a single `SubmittedMetadata`.

> 1. These are then appended to lists for batching, one `SubmittedRollupDataList`

Reviewer: ```suggestion
2. These are then appended to lists for batching, one `SubmittedRollupDataList`
```

  per rollup (a collection of `SubmittedRollupData`s), and a
  `SubmittedMetadataList` (a collection of `SubmittedMetadata`s). At this stage,
  filtering of rollups is applied if enabled: only rollups specified in the
  `ASTRIA_SEQUENCER_RELAYER_ONLY_INCLUDE_ROLLUPS` env var are included, or else
  no filtering is done if the env var is empty.

> 1. All the lists are converted to a single `Payload` of Celestia blobs, one list

Reviewer: ```suggestion
3. All the lists are converted to a single `Payload` of Celestia blobs, one list
```

  per blob. Each list is encoded to bytes using Protobuf serialization then
  compressed using Brotli. The payload also contains the total uncompressed and
  compressed size of the data (specifically the number of bytes in the
  Protobuf-encoded data before and after compression).

Any error encountered during data transformation is cause for the entire process
to exit, with the exception of exceeding the batch size limit, where the last
block is instead cached until the current batch has been successfully put to
Celestia.

### Putting the Data onto Celestia

Submitting the data in the form of Celestia blobs is part of the submitter task,
and involves several steps where it communicates with a single Celestia app via
gRPC as follows:

> 1. Four RPCs are sent in parallel to retrieve the relayer's Celestia account

Reviewer: also fix list numbers here
